File: agentic-patterns/js/src/autonomous-operation.ts
import { z } from 'genkit';
import { ai } from './genkit.js';
import { googleAI } from '@genkit-ai/google-genai';

// A tool for the agent to search the web
const searchWeb = ai.defineTool(
  {
    name: 'searchWeb',
    description: 'Search the web for information on a given topic.',
    inputSchema: z.object({ query: z.string() }),
    outputSchema: z.string(),
  },
  async ({ query }) => {
    // In a real app, you would implement a web search API call here.
    return `You found search results for: ${query}`;
  }
);

// A tool for the agent to ask the user a question
const askUser = ai.defineInterrupt(
  {
    name: 'askUser',
    description: 'Ask the user a clarifying question.',
    inputSchema: z.object({ question: z.string() }),
    outputSchema: z.string(),
  },
);

export const researchAgent = ai.defineFlow(
  {
    name: 'researchAgent',
    inputSchema: z.object({ task: z.string() }),
    outputSchema: z.string(),
  },
  async ({ task }) => {
    let response = await ai.generate({
      system: `You are a helpful research assistant. Your goal is to provide a comprehensive answer to the user's task.`,
      prompt: `Your task is: ${task}. Use the available tools to accomplish this.`,
      model: googleAI.model('gemini-2.5-pro'),
      tools: [searchWeb, askUser],
      maxTurns: 5, // Limit the number of back-and-forth turns
    });

    // Handle potential interrupts (e.g., asking the user a question)
    while (response.interrupts.length > 0) {
      const interrupt = response.interrupts[0];
      if (interrupt.toolRequest.name === 'askUser') {
        const question = (interrupt.toolRequest.input as any).question;
        
        // In a real app, you would present the question to the user and get their answer.
        const userAnswer = await Promise.resolve(`The user answered: "Sample answer for '${question}'"`);

        response = await ai.generate({
          messages: response.messages,
          tools: [searchWeb, askUser],
          resume: {
            respond: [askUser.respond(interrupt, userAnswer)],
          },
        });
      } else {
        // Handle other unexpected interrupts if necessary
        break;
      }
    }

    return response.text;
  }
);

File: agentic-patterns/dart/lib/autonomous_operation.dart
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'package:genkit/genkit.dart';
import 'package:schemantic/schemantic.dart';

part 'autonomous_operation.g.dart';

@Schema()
abstract class $AutonomousOperationInput {
  @Field(defaultValue: 'Research the current state of Genkit Dart support.')
  String get goal;
}

@Schema()
abstract class $AgentSearchInput {
  String get query;
}

@Schema()
abstract class $AgentAskUserInput {
  String get question;
}

void defineAutonomousOperationFlows(Genkit ai) {
  // 1. Define tools for the agent
  final webSearch = ai.defineTool(
    name: 'webSearch',
    description: 'Search the web for information.',
    inputSchema: AgentSearchInput.$schema,
    outputSchema: .string(),
    fn: (input, _) async {
      // In a real app, you would call a search API here.
      return 'Search results for "${input.query}": Genkit is a powerful AI framework from Firebase. '
          'This tool is mocked for demonstration purposes. In a real app, you would call a search API here. '
          'For now, pretend you found relevant information.';
    },
  );

  final askUser = ai.defineTool(
    name: 'askUser',
    description: 'Ask the user for clarification or more information.',
    inputSchema: AgentAskUserInput.$schema,
    outputSchema: .string(),
    fn: (input, context) async {
      // This tool interrupts the flow to ask the user a question.
      context.interrupt(input.question);
    },
  );

  // 2. Define the autonomous agent flow
  ai.defineFlow(
    name: 'researchAgent',
    inputSchema: AutonomousOperationInput.$schema,
    outputSchema: .string(),
    fn: (input, _) async {
      var response = await ai.generate(
        messages: [
          Message(
            role: Role.system,
            content: [
              TextPart(
                text:
                    'You are a research agent. Your goal is to help the user with their research goal. '
                    'Use the provided tools to search the web and ask the user for more information if needed. '
                    'Plan your steps and execute them autonomously.',
              ),
            ],
          ),
        ],
        prompt: input.goal,
        toolNames: [webSearch.name, askUser.name],
      );

      // Handle potential interrupts (human-in-the-loop)
      while (response.finishReason == FinishReason.interrupted) {
        final interrupts = response.interrupts;
        if (interrupts.isEmpty) break;

        final resumeResponses = <InterruptResponse>[];
        for (final interrupt in interrupts) {
          if (interrupt.toolRequest.name == 'askUser') {
            final question = interrupt.metadata?['interrupt'] as String?;
            // In a real app, you'd prompt the user here. For this sample:
            final simulatedAnswer =
                'The user answered: "Sample answer for \'$question\'"';
            resumeResponses.add(
              InterruptResponse(interrupt.toolRequestPart!, simulatedAnswer),
            );
          }
        }

        response = await ai.generate(
          messages: response.messages,
          toolNames: [webSearch.name, askUser.name],
          interruptRespond: resumeResponses,
        );
      }

      return response.text;
    },
  );
}
